### ISSUE TITLE
TestCase for CodeBuilder.defaults property

### ISSUE THREAD

**Issue by blaggacao** [Fatal1ty/mashumaro#14]
It seems to me as if the `defaults` property of the `CodeBuilder` is nurtured by
`self.cls.__dict__` (via `namepsace` property).

If I look at an actual class I got:
`cls.__dataclass_fields__` holding a mapping from field names to dataclass Field instances, which in turn has either a `default` or a `default_factory`.

Maybe the default property should read values from the `cls.__dataclass_fields__` just as normally instanciated dataclass? (or at least not raise a MissingError, as `cls(**kwargs)` would fill them in afterwards, and only break if default is a `NoneType`)

**Comment by blaggacao** [comment]
Maybe I did not understand `defaults` properly, hence asking for a test case :wink:

**Comment by Fatal1ty** [comment]
The problem is `__dataclass_fields__` property doesn't exist until we created the class because the property is set inside `dataclass` decorator.

**Comment by Fatal1ty** [comment]
So it will be a problem when one data class is inherited from another data class that have default values.

```python
@dataclass
class A:
    a: int
    b: int = 123

@dataclass
class B(A, DataClassDictMixin):
    c: int = 456

B.from_dict({'a': 111, 'c': 222})  # mashumaro.exceptions.MissingField: Field "b" of type builtins.int is missing in __main__.B instance
```

**Comment by blaggacao** [comment]
Looks like the abstract example you give is what I'm experiencing. As a workaround I dummy-initialized the dataclass with default values, but this required me to put an obscure validation `InitArg` to pass to the `__post_init__` to mute the data validator as defaults fore the sole purpose of seeding missing values might not be valid values in some circumestances and rather might been overridden from the actual dict for the final dataclass object.

**Comment by blaggacao** [comment]
```python

class BaseConfig(DataClassDictMixin):

    @classmethod
    def load(cls, confd: os.PathLike):
        cls.confd = confd
        # Dummy instanciation as mashumaro doesn't support defaulting
        # see: https://github.com/Fatal1ty/mashumaro/issues/14
        default_instance = cls(validate=False)
        cfg = {}
        cfg = default_instance.to_dict()
        for child in cls.confd.iterdir():
            if child.name == cls._file_name:
                try:
                    cfg.update(json.loads(child.read_text()))
                except json.decoder.JSONDecodeError as e:
                    _log.critical(f"File {child} does not contain vaild json.")
                    raise e
                break
        cls._cure(cfg)
        try:
            return cls.from_dict(cfg)
        except TypeError as e:
            raise InvalidOptionError(e.args[0])

    def __post_init__(self, validate):
        if not validate:
            return
        self._validate()


@dataclass(frozen=True)
class OdooConfig(BaseConfig):

    validate: InitVar[bool] = True
    ...


@dataclass(frozen=True)
class OdooConfigProd(OdooConfig):
    _file_name = "odooconfig.prod.json"

```

**Comment by Fatal1ty** [comment]
Hi there. I managed to fix this in 1.11, so @blaggacao you don't need to have a workaround anymore. Can you check if everything seems to be ok?

**Comment by blaggacao** [comment]
Nice! I'll check asap... I'm still in the sprint, so thanks tons for your responsiveness. I'll report back.

**Comment by blaggacao** [comment]
I spotted an issue if the inherited class does not define any annotations (e.g. only class variables, but without typing them `as typing.ClassVar`).
I'm trying to put together a test case, and probably a fix...

**Comment by Fatal1ty** [comment]
@blaggacao thank you. I made the quick in the published 1.12 release.

**Comment by blaggacao** [comment]
:heavy_check_mark:
